### PR TITLE
feature: support self-attention guidance with SD1 inpainting model

### PR DESCRIPTION
**Pull request by brycedrennan**
This is basically the change, which adds the mask and target image latents when inputing to the unet and then discards them.  I have not done extensive comparisons to validate the efficacy of this but it appears to work.

I also wonder if the effect would be improved if the self-attention was focused on just the area being inpainted. 

```python
        x = torch.cat(
            tensors=(degraded_latents, self.mask_latents, self.target_image_latents),
            dim=1,
        )
        result = self.unet(x)
        degraded_noise = result[:, :4, :, :]
```
Comparison
animated_with_text
animated_with_text_new

Inpainting without Self-Attention Guidance
test_inpainting_bench_

Inpainting with Self-Attention Guidance
test_inpainting_bench_with_sag

Original:
test_inpainting_bench__orig

### CODE REVIEW

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py [hunk 1]
```diff
@@ -143,3 +143,27 @@ def set_inpainting_conditions(
...
+
+        negative_embedding, _ = clip_text_embedding.chunk(2)
+        timestep = self.scheduler.timesteps[step].unsqueeze(dim=0)
+        self.set_unet_context(timestep=timestep, clip_text_embedding=negative_embedding, **kwargs)
+        x = torch.cat(
+            tensors=(degraded_latents, self.mask_latents, self.target_image_latents),
+            dim=1,
+        )
+        result = self.unet(x)
+        degraded_noise = result[:, :4, :, :]
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py:167]: This is not needed: the predicted noise has just 4 channels

(only the first conv of the unet is impacted - 4 -> 9 channels - the rest remains identical)

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py [hunk 2]
```diff
@@ -143,3 +143,27 @@ def set_inpainting_conditions(
         self.target_image_latents = self.lda.encode(x=masked_init_image)
 
         return self.mask_latents, self.target_image_latents
+
+    def compute_self_attention_guidance(self, x, noise, step: int, *, clip_text_embedding, **kwargs):
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py:147]: type hints are missing

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py [hunk 3]
```diff
@@ -143,3 +143,27 @@ def set_inpainting_conditions(
...
+            latents=x,
+            noise=noise,
+            step=step,
+            classifier_free_guidance=True,
+        )
+
+        negative_embedding, _ = clip_text_embedding.chunk(2)
+        timestep = self.scheduler.timesteps[step].unsqueeze(dim=0)
+        self.set_unet_context(timestep=timestep, clip_text_embedding=negative_embedding, **kwargs)
+        x = torch.cat(
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py:162]: To please the type checker, you'll need additional asserts like [these](https://github.com/finegrain-ai/refiners/blob/ab0915d0525523bfb5e33d65acd584275888052e/src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py#L118-L119)

**Review by deltheil — COMMENTED** [review]
Thanks!

**Comment by brycedrennan** [comment]
Great feedback.  All addressed now.

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py [hunk 4]
```diff
@@ -143,3 +143,30 @@ def set_inpainting_conditions(
         self.target_image_latents = self.lda.encode(x=masked_init_image)
 
         return self.mask_latents, self.target_image_latents
+
+    def compute_self_attention_guidance(
+        self, x: Tensor, noise: Tensor, step: int, *, clip_text_embedding: Tensor, **kwargs: Tensor
+    ):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py:149]:
```suggestion
    ) -> Tensor:
```
little nit

**Review by limiteinductive — COMMENTED** [review]
This seems good to me 👍

File: src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py (hunk 4 above)
Comment (brycedrennan) [reply src/refiners/foundationals/latent_diffusion/stable_diffusion_1/model.py:149]: ah yes i missed that

**Review by deltheil — APPROVED** [review]
